execution_engine_testing/tests/src/test/regression/gh_3097.rs

```diff
@@ -0,0 +1,354 @@
+        GH_3097_REGRESSION_CALL_WASM,
+        runtime_args! {
+            ARG_METHOD => METHOD_CALL_VERSIONED_CONTRACT,
+            ARG_CONTRACT_VERSION => Some(1u32),
+        },
+    )
+    .build();
+
+    let session_call_v2_request = ExecuteRequestBuilder::standard(
+        *DEFAULT_ACCOUNT_ADDR,
+        GH_3097_REGRESSION_CALL_WASM,
+        runtime_args! {
+            ARG_METHOD => METHOD_CALL_VERSIONED_CONTRACT,
+            ARG_CONTRACT_VERSION => Some(2u32),
+        },
+    )
+    .build();
+
+    let session_call_latest_request = ExecuteRequestBuilder::standard(
+        *DEFAULT_ACCOUNT_ADDR,
+        GH_3097_REGRESSION_CALL_WASM,
+        runtime_args! {
+            ARG_METHOD => METHOD_CALL_VERSIONED_CONTRACT,
+            ARG_CONTRACT_VERSION => Option::<u32>::None,
+        },
+    )
+    .build();
+
+    builder
+        .exec(session_call_latest_request)
+        .expect_success()
+        .commit();
+
+    builder
+        .exec(session_call_v2_request)
+        .expect_success()
+        .commit();
+
+    builder
+        .exec(session_call_v1_request)
+        .expect_failure()
+        .commit();
+
+    let error = builder.get_error().expect("should have error");
+    assert!(
+        matches!(
+            error,
+            casper_execution_engine::core::engine_state::Error::Exec(
+                casper_execution_engine::core::execution::Error::InvalidContractVersion(version)
+            )
+            if version == ContractVersionKey::new(1, 1),
+        ),
+        "Expected invalid contract version, found {:?}",
+        error,
+    );
+
+    // Call by contract hashes
+
+    let call_by_hash_v2_request = ExecuteRequestBuilder::contract_call_by_hash(
+        *DEFAULT_ACCOUNT_ADDR,
+        contract_hash_v2,
+        DO_SOMETHING_ENTRYPOINT,
+        RuntimeArgs::new(),
+    )
+    .build();
+
+    builder
+        .exec(call_by_hash_v2_request)
+        .expect_success()
+        .commit();
+
+    let call_by_name_v2_request = ExecuteRequestBuilder::contract_call_by_name(
+        *DEFAULT_ACCOUNT_ADDR,
+        CONTRACT_HASH_V2_KEY,
+        DO_SOMETHING_ENTRYPOINT,
+        RuntimeArgs::new(),
+    )
+    .build();
+    builder
+        .exec(call_by_name_v2_request)
+        .expect_success()
+        .commit();
+
+    // This direct contract by name/hash should fail
+    let call_by_hash_v1_request = ExecuteRequestBuilder::contract_call_by_hash(
+        *DEFAULT_ACCOUNT_ADDR,
+        contract_hash_v1,
+        DO_SOMETHING_ENTRYPOINT,
+        RuntimeArgs::new(),
+    )
+    .build();
+    builder
+        .exec(call_by_hash_v1_request)
+        .expect_failure()
+        .commit();
+
+    let error = builder.get_error().expect("should have error");
+    assert!(
+        matches!(
+            error,
+            casper_execution_engine::core::engine_state::Error::Exec(
+                casper_execution_engine::core::execution::Error::DisabledContract(disabled_contract_hash)
```

Comment:
The test looks ok, but I had some trouble understanding why we expect this contract to be disabled. This is (probably) because earlier we executed the other contract that called `storage::disable_contract_version()` on `contract_hash_v1`.

Can you maybe consider adding more comments that explain the test flow for people not that well acquainted with the execution engine?

Reply:
Fair point. I renamed `contract_hash_v1`  and _v2 into `enabled_contract_hash` and `disabled_contract_hash` to make this a bit more readable. I hope that helps with readability